### PR TITLE
🌱 Bump CAPI to v1.6, Helm to v3.13.3, and controller-runtime to v0.16.3

### PR DESCRIPTION
**Pull request by Jont828**
**What this PR does / why we need it**: Bump CAPI version. This should unblock importing the CAPI test framework as previous versions of CAPI test was incompatible with Helm. Doing so will allow us to implement a e2e test suite.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -576,8 +577,6 @@ set-manifest-image:
 .PHONY: clean
 clean: ## Remove generated binaries, GitBook files, Helm charts, and Tilt build files
 	$(MAKE) clean-bin
-	$(MAKE) clean-charts
```
Comment (mboersma) [inline Makefile:579]: Looks like these targets didn't exist, so I assume this is just cleanup?

File: go.mod [hunk 2]
```diff
@@ -1,41 +1,25 @@
...
-	sigs.k8s.io/cluster-api v1.5.3
-	sigs.k8s.io/controller-runtime v0.15.1
+	helm.sh/helm/v3 v3.13.3
+	k8s.io/api v0.28.4
+	k8s.io/apimachinery v0.28.4
+	k8s.io/cli-runtime v0.28.4
+	k8s.io/client-go v0.28.4
+	k8s.io/klog/v2 v2.100.1
+	k8s.io/utils v0.0.0-20240102154912-e7106e64919e
+	sigs.k8s.io/cluster-api v1.6.0
```
Comment (mboersma) [inline go.mod:21]: Should we bump this up to [CAPI v1.6.1](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.6.1) now?

File: main.go [hunk 3]
```diff
@@ -50,53 +70,99 @@ func init() {
...
 		"Duration the LeaderElector clients should wait between tries of actions (duration string)")
-	flag.IntVar(&helmChartProxyConcurrency, "helm-chart-proxy-concurrency", 10, "The number of HelmChartProxies to process concurrently.")
-	flag.IntVar(&helmReleaseProxyConcurrency, "helm-release-proxy-concurrency", 10, "The number of HelmReleaseProxies to process concurrently.")
-	flag.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
+
+	fs.StringVar(&watchNamespace, "namespace", "",
+		"Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.")
+
+	fs.StringVar(&watchFilterValue, "watch-filter", "",
+		fmt.Sprintf("Label value that the controller watches to reconcile cluster-api objects. Label key is always %s. If unspecified, the controller watches for all cluster-api objects.", clusterv1.WatchLabel))
```
Comment (mboersma) [inline main.go:91]:
```suggestion
		fmt.Sprintf("Label value that the controller watches to reconcile cluster-api objects. The label key is always %s. If unspecified, the controller watches for all cluster-api objects.", clusterv1.WatchLabel))
```

File: main.go [hunk 4]
```diff
@@ -50,53 +70,99 @@ func init() {
...
+
+	fs.StringVar(&profilerAddress, "profiler-address", "",
+		"Bind address to expose the pprof profiler (e.g. localhost:6060)")
+
+	fs.IntVar(&helmChartProxyConcurrency, "helm-chart-proxy-concurrency", 10,
+		"The number of HelmChartProxies to process concurrently.")
```
Comment (mboersma) [inline main.go:97]:
```suggestion
		"Number of HelmChartProxies to process concurrently.")
```

Just for consistency; it seems that most of the flag descriptions start without "The".

File: main.go [hunk 5]
```diff
@@ -50,53 +70,99 @@ func init() {
...
+
+	fs.IntVar(&helmReleaseProxyConcurrency, "helm-release-proxy-concurrency", 10,
+		"The number of HelmReleaseProxies to process concurrently.")
```
Comment (mboersma) [inline main.go:100]:
```suggestion
		"Number of HelmReleaseProxies to process concurrently.")
```

File: main.go [hunk 6]
```diff
@@ -50,53 +70,99 @@ func init() {
...
+
+	fs.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
 		"The minimum interval at which watched resources are reconciled (e.g. 15m)")
```
Comment (mboersma) [inline main.go:103]:
```suggestion
		"Minimum interval at which watched resources are reconciled (e.g. 15m)")
```

File: main.go [hunk 7]
```diff
@@ -50,53 +70,99 @@ func init() {
...
+
+	fs.IntVar(&webhookPort, "webhook-port", 9443,
+		"Webhook Server port")
+
+	fs.StringVar(&webhookCertDir, "webhook-cert-dir", "/tmp/k8s-webhook-server/serving-certs/",
+		"Webhook cert dir, only used when webhook-port is specified.")
+
+	fs.StringVar(&healthAddr, "health-addr", ":9440",
+		"The address the health endpoint binds to.")
```
Comment (mboersma) [inline main.go:112]:
```suggestion
		"Address the health endpoint binds to.")
```

File: main.go [hunk 8]
```diff
@@ -50,53 +70,99 @@ func init() {
...
 
-	ctrl.SetLogger(klogr.NewWithOptions(klogr.WithFormat(klogr.FormatKlog)))
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme:                 scheme,
-		MetricsBindAddress:     metricsAddr,
-		Port:                   9443,
-		HealthProbeBindAddress: probeAddr,
 		LeaderElection:         enableLeaderElection,
-		LeaderElectionID:       "5a2dee3e.cluster.x-k8s.io",
+		LeaderElectionID:       "controller-leader-election-caaph",
```
Comment (mboersma) [inline main.go:149]: This seems better. :-) Is there any convention for `LeaderElectionID` in other controllers? (Should it include a namespace like `cluster.x-k8s.io`?)

**Review by mboersma — COMMENTED** [review]
This lgtm, I just had some minor comments.

File: go.mod (hunk 2 above)
Comment (Jont828) [reply go.mod:21]: I'm in favor of leaving it for now and bumping it once we merge the e2e test framework PR so we don't need to manually verify that a dependency bump doesn't break anything.

File: main.go (hunk 8 above)
Comment (Jont828) [reply main.go:149]: I copied this from CAPZ, which had it as `controller-leader-election-capz`. Slowly working through the copy/paste issues!

**Comment by Jont828** [comment]
@mboersma Rebased and squashed!

File: Makefile [hunk 9]
```diff
@@ -134,17 +134,17 @@ ENVSUBST_VER := $(call get_go_version,github.com/drone/envsubst/v2)
...
 GO_APIDIFF_PKG := github.com/joelanford/go-apidiff
 
-HADOLINT_VER := v2.10.0
+HADOLINT_VER := v2.12.0
 HADOLINT_FAILURE_THRESHOLD = warning
 
 SHELLCHECK_VER := v0.9.0
 
-KPROMO_VER := v3.5.1
+KPROMO_VER := v4.0.4
```
Comment (mboersma) [inline Makefile:147]: You may need to update a line farther down in the file to `v4`; I made this mistake in CAPZ and here's my fix: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4478

Comment (mboersma) [reply Makefile:147]: https://github.com/kubernetes-sigs/cluster-api-addon-provider-helm/blob/8d7fcb998f1224b0fb6511ea5f127adeb1a008c0/Makefile#L150

Should end with `v4/cmd/kpromo`.

Comment (Jont828) [reply Makefile:147]: Nice catch!

**Review by mboersma — APPROVED** [review]
/lgtm

**Pull request merged**